contracts/halo-pool/src/contract.rs

```diff
@@ -18,46 +20,51 @@ use crate::{
     formulas::{calc_reward_amount, get_multiplier, update_pool},
     msg::{ExecuteMsg, InstantiateMsg, QueryMsg},
     state::{
-        Config, PoolInfo, RewardTokenAsset, StakerRewardAssetInfo, TokenInfo,
-        ACCRUED_TOKEN_PER_SHARE, CONFIG, LAST_REWARD_TIME, POOL_INFO, STAKERS_INFO,
+        Config, PoolInfo, PoolInfos, RewardTokenAsset, RewardTokenAssetResponse,
+        StakerInfoResponse, TokenInfo, CONFIG, POOL_INFOS, STAKERS_INFO,
     },
 };
 
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn instantiate(
     deps: DepsMut,
     _env: Env,
-    info: MessageInfo,
+    _info: MessageInfo,
     msg: InstantiateMsg,
 ) -> StdResult<Response> {
     set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
 
     let config = Config {
-        halo_factory_owner: info.sender,
+        halo_factory_owner: msg.pool_owner,
     };
 
-    let pool_info = &PoolInfo {
-        staked_token: deps.api.addr_validate(&msg.staked_token)?.to_string(),
-        reward_token: msg.reward_token.clone(),
-        reward_per_second: Decimal::zero(), // will be updated when admin adding reward balance
+    // Init pool info
+    let pool_info = PoolInfo {
+        reward_per_second: Decimal::zero(), // will be updated after admin adding reward balance
         start_time: msg.start_time,
         end_time: msg.end_time,
         pool_limit_per_user: msg.pool_limit_per_user,
-        whitelist: msg.whitelist,
+        total_staked_at_end_time: Uint128::zero(),
     };
 
+    // Init pool infos
+    POOL_INFOS.save(
+        deps.storage,
+        &PoolInfos {
+            staked_token: deps.api.addr_validate(&msg.staked_token)?.to_string(),
```

Comment:
Chuyển kiểu dữ liệu của thằng này thành Addr đi. Từ String, test sang Addr rồi lại quay về String hơi kỳ

Reply:
Updated to Add